### PR TITLE
plumb HashMeta through sigchains and into key family

### PR DESCRIPTION
**Pull request by maxtaco**
- fix a logstanding bug with EldestKey delegation, which wasn't getting Delegate() called on in it when walking the chain
  - as a result, we couldn't revoke an Eldest by a sigID -- it would do something crazy
  - introduce an EldestChainLink type to fix it
  - plumb through all differences
- Also introduce EldestSeqno and FirstAppearedMerkleSeqnoUnverified into chain links
  - not being used yet, but mabye in the future
- Introduce new deepcopy upgrade for node avdl compiler

### CODE REVIEW

**Comment by keybase-ci-visdiff** [comment]
The commits 1ed57ad12100...bd2c99f4b7f9 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Markdown-emoji** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496807286887-1ed57ad12100-bd2c99f4b7f9-linux/Markdown-emoji-1ed57ad12100.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1496807286887-1ed57ad12100-bd2c99f4b7f9-linux/Markdown-emoji-bd2c99f4b7f9.png)  
   Markdown-emoji comparison
</summary>

**Comment by maxtaco** [comment]
@oconnor663  or @mlsteele, whoever can grab it before sprint close!

File: go/libkb/chain_link.go [hunk 1]
```diff
@@ -490,6 +515,14 @@ func (c *ChainLink) Unpack(trusted bool, selfUID keybase1.UID) (err error) {
 	if err != nil {
 		return err
 	}
+
+	if i, err := c.packed.AtKey("merkle_seqno").GetInt64(); err == nil {
+		tmp.firstAppearedMerkleSeqnoUnverified = keybase1.Seqno(i)
+	}
+	if i, err := c.packed.AtKey("eldest_seqno").GetInt64(); err == nil {
+		tmp.eldestSeqno = keybase1.Seqno(i)
```
Comment (oconnor663) [inline go/libkb/chain_link.go]: Are we trusting this value from the server?

File: go/libkb/sig_chain.go [hunk 2]
```diff
@@ -83,7 +83,9 @@ func (c ChainLinks) EldestSeqno() keybase1.Seqno {
 	return c[0].GetSeqno()
 }
 
-func (sc *SigChain) LocalDelegate(kf *KeyFamily, key GenericKey, sigID keybase1.SigID, signingKid keybase1.KID, isSibkey bool) (err error) {
+func (sc *SigChain) LocalDelegate(kf *KeyFamily, key GenericKey, sigID keybase1.SigID, signingKid keybase1.KID, isSibkey bool, mhm keybase1.HashMeta, fau keybase1.Seqno) (err error) {
```
Comment (oconnor663) [inline go/libkb/sig_chain.go]: What do `mhm` and `fau` stand for?

**Comment by maxtaco** [comment]
Yeah though we actually aren't using it right now

**Comment by oconnor663** [comment]
Any reason not to just compute it locally when we walk the chain? Is this complicated by using the same function to unpack links from the server and to unpack links form local storage?

File: go/libkb/sig_chain.go (hunk 2 above)
Comment (maxtaco) [inline go/libkb/sig_chain.go]: abbrvtns for lng vrbl nms

**Comment by maxtaco** [comment]
What's "it"?

**Comment by oconnor663** [comment]
Yes, `eldest_seqno`. That's what you meant when you said "we aren't using it right now", yes?

File: go/libkb/id_table.go [hunk 3]
```diff
@@ -1092,8 +1132,8 @@ func NewTypedChainLink(cl *ChainLink) (ret TypedChainLink, w Warning) {
 		err = fmt.Errorf("No type in signature @%s", base.ToDebugString())
 	} else {
 		switch s {
-		case "eldest":
-			ret, err = ParseSelfSigChainLink(base)
```
Comment (oconnor663) [inline go/libkb/id_table.go]: Should we delete the SelfSigChainLink type now?

**Comment by maxtaco** [comment]
yes, sorry, context was lost when i emailed you from the train!

File: go/libkb/id_table.go (hunk 3 above)
Comment (oconnor663) [inline go/libkb/id_table.go]: Ah never mind, `ParseWebServiceBinding` uses it.

File: go/libkb/keyfamily.go [hunk 4]
```diff
@@ -562,11 +583,15 @@ func (ckf *ComputedKeyFamily) DelegatePerUserKey(perUserKey keybase1.PerUserKey)
...
+// fau = "FirstAppearedUnverified", a hint from the server that we're going to persist.
+func (cki *ComputedKeyInfos) Delegate(kid keybase1.KID, tm *KeybaseTime, sigid keybase1.SigID, signingKid, parentKID keybase1.KID,
+	pgpHash string, isSibkey bool, ctime, etime time.Time,
+	merkleHashMeta keybase1.HashMeta, fau keybase1.Seqno) (err error) {
+
+	cki.G().Log.Debug("ComputeKeyInfos#Delegate To %s with %s at sig %s", kid.String(), signingKid, sigid.ToDisplayString(true))
 	info, found := cki.Infos[kid]
 	if !found {
-		newInfo := NewComputedKeyInfo(false, false, KeyUncancelled, ctime.Unix(), etime.Unix(), pgpHash)
+		newInfo := NewComputedKeyInfo(kid, false, false, KeyUncancelled, ctime.Unix(), etime.Unix(), pgpHash)
```
Comment (oconnor663) [inline go/libkb/keyfamily.go]: Should we just pass the `isSibkey` parameter through here? Looks like we set it below, but it feels weird to supply the wrong value at first.

File: go/libkb/chain_link.go (hunk 1 above)
Comment (maxtaco) [inline go/libkb/chain_link.go]: ok gone

File: go/libkb/keyfamily.go (hunk 4 above)
Comment (maxtaco) [inline go/libkb/keyfamily.go]: ok

**Pull request merged**